Feature/frontend layout

norling opened the pull request:
It's not pretty, but I'm hoping that it's enough functionality to test during the summer! :)

norling committed:
make string trimming more understandable

norling committed:
test: update tests to new permission system

norling committed:
Add material tab menu

norling committed:
Add material tab menu

norling committed:
Add material tab menu

norling committed:
Improve tab menu routing

norling committed:
Improve tab menu routing

norling committed:
Improve tab menu routing

norling committed:
Add basic managment page

norling committed:
Add basic managment page

norling committed:
Add basic managment page

norling committed:
Make login form look nicer

norling committed:
Make login form look nicer

norling committed:
Make login form look nicer

norling committed:
Add main div to content

norling committed:
Add main div to content

norling committed:
Add main div to content

Commit:
Fake onSubmit when pressing enter on Login form

Commit:
Fake onSubmit when pressing enter on Login form

Commit:
Fake onSubmit when pressing enter on Login form

Commit:
Activate correct tab when location changes

Commit:
Activate correct tab when location changes

Commit:
Activate correct tab when location changes

Commit:
Explicit state changes in user context

Commit:
Explicit state changes in user context

Commit:
Explicit state changes in user context

Commit:
Rearrange tabs code for readability

Commit:
Rearrange tabs code for readability

Commit:
Rearrange tabs code for readability

Commit:
Add half-finished attempt at getting tests to work

Commit:
Add half-finished attempt at getting tests to work

Commit:
Add half-finished attempt at getting tests to work

norling committed:
Add tab structure for manage page

norling committed:
Add tab structure for manage page

norling committed:
Add tab structure for manage page

norling committed:
Merge pull request #74 from NBISweden/danr/frontend-layout

Clean up suggestion of the new front-end code

norling committed:
Merge branch 'feature/frontend-layout' of github.com:NBISweden/theherdbook into feature/frontend-layout

norling committed:
Exclude /dist/* in tsconfig.json

norling committed:
Exclude /dist/* in tsconfig.json

norling committed:
fix: Check before using genebank value

norling committed:
fix: Check before using genebank value

norling committed:
Add /api/users endpoint

norling committed:
Add /api/users endpoint

norling committed:
Add update_role() for managing user permissions

norling committed:
Add update_role() for managing user permissions

norling committed:
Weird css calc() behavior

norling committed:
Weird css calc() behavior

norling committed:
clean css to make manage page fixed height

norling committed:
clean css to make manage page fixed height

norling committed:
add basic herd management form

norling committed:
add basic herd management form

norling committed:
test: test endpoints without running docker

norling committed:
test: test endpoints without running docker

norling committed:
test for update_role

norling committed:
test for update_role

norling committed:
test: improved get_user tests

norling committed:
test: improved get_user tests

norling committed:
fix: make get_users return proper json

norling committed:
fix: make get_users return proper json

norling committed:
test: Add tests for get_users endpoint

norling committed:
test: Add tests for get_users endpoint

norling committed:
Separate herd form into own file

norling committed:
Separate herd form into own file

norling committed:
Add controls for updating basic herd data

norling committed:
Add controls for updating basic herd data

norling committed:
Add user selection tabs to manage page

norling committed:
Add user selection tabs to manage page

norling committed:
Basic user management form

norling committed:
Basic user management form

norling committed:
Merge branch 'feature/frontend-layout' of github.com:NBISweden/theherdbook into feature/frontend-layout

norling committed:
Add remove role button

norling committed:
data setting problem

Commit:
Put the DataContext global in its own file to help HMR

norling committed:
Data context working with genebanks

norling committed:
Add herd list to data context

norling committed:
Split data access functions from database file

norling committed:
Move user list to data context

norling committed:
Add "add user" button

norling committed:
subrouting debug commit

norling committed:
fix: update import in register_user

norling committed:
update genebank page including individuals

norling committed:
Add basic material table to individuals page

Commit:
Rework all routing and tabbing

norling committed:
Merge pull request #80 from NBISweden/danr/routing

Rework all routing and tabbing

norling committed:
Update genebanks page to use new routing

norling committed:
Add material-table for individuals table

norling committed:
fix: only load herds if there are herds

Commit:
Silence the warning about no tab selected

norling committed:
Merge pull request #81 from NBISweden/fix/no-tab-selected

Silence the warning about no tab selected

norling committed:
fix: don't request when id is undefined

norling committed:
fix: add <!doctype html> to index.html

norling committed:
Add "add user" functionality

norling committed:
Hide "add herd" button

norling committed:
fix: Allow owner role operations

norling committed:
Add form for adding user roles

norling committed:
css: height fixes to keep things in window

norling committed:
Treat is_active as True for selecting herds

norling committed:
fix: herd name in manage list

norling committed:
Allow herd.is_active None to select individuals

norling committed:
pylint clean-up

pontus commented on app/utils/database.py:
```diff
@@ -462,6 +531,30 @@ def is_admin(self):
         return 'admin' in [p['level'] for p in self.privileges]
 
+    @property
+    def is_manager(self):
```
Can we get a little reasoning (in PR is fine) about the naming convention here; why is it is_manager (which implies booleanity to me) and not genebanks_managed (or preferably a better name)?

norling replied:
yeah... it's named that way since it used to be a boolean ^^ Then I changed it to give extra info. I should separate the two though!

danr commented on frontend/genebanks.tsx:
```diff
@@ -1,33 +1,98 @@
+
+import * as ui from './ui_utils'
+
+function InnerPaper(props: {id: number}) {
```
Oops I had hoped that you would find a more descriptive name than `InnerPaper` here :D

norling replied:
it works for now ^^

danr commented on frontend/herdForm.tsx:
```diff
@@ -0,0 +1,112 @@
+          {simpleFields.map((field: any, i: number) => {
+            return <TextField
+                    key={i}
+                    id={field.key}
```
Rename key to id then ;)

norling replied:
I prefer it as 'key' or 'name' as it's a reference to key->value pairs of the herd object. I can remove the id tag though.

danr replied:
I see! I got confused thinking it would be react keys.

danr commented on frontend/herdForm.tsx:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  const submitForm = () => {
+    let postData = Object.assign({}, herd);
```
`const postData = {...herd}`

norling replied:
I don't mind this syntax, but you can change to spread if you like ^^

danr replied:
OK! :grimacing: 

But please use `const` when you don't need `let`!

danr commented on frontend/herdForm.tsx:
```diff
@@ -0,0 +1,112 @@
+/**
+ * Provides herd management forms for setting herd metadata.
+ */
+export function HerdForm(props: {id: any | undefined}) {
```
What's with the type? `any | undefined` = `any`. Did you mean `string | undefined`?

norling replied:
haha - I think I meant string | number | undefined, but got confused ^^

danr commented on frontend/herdForm.tsx:
```diff
@@ -0,0 +1,112 @@
+ * Provides herd management forms for setting herd metadata.
+ */
+export function HerdForm(props: {id: any | undefined}) {
+  const [herd, setHerd] = React.useState(undefined as any)
```
You have types for herds, why not use `undefined | Herd`?

norling replied:
yeah, that was the plan - I just ran out of time, so the old code never got the new types. I'll have a look now while we're looking!

danr commented on frontend/herdForm.tsx:
```diff
@@ -0,0 +1,112 @@
+
+  const setFormField = (label: string, value: string | number) => {
+    herd[label] = value;
+    setHerd(herd);
```
`herd && setHerd({...herd, [label]: value})`

danr replied:
Also you can type `const setFormField = <K extends keyof Herd>(label: K, value: Herd[K]) => ...` :nerd_face:

norling replied:
haha - I considered something like that but didn't get it to work, but I'll use this ^^

danr commented on frontend/herdForm.tsx:
```diff
@@ -0,0 +1,112 @@
+        </Button>
+        <h2>Individer</h2>
+        <ul>
+          {herd.individuals.map((individual: any, i:number) => {
```
Hmm `Herd`s have no `individuals`... Is this actually a `Genebank`? :thinking:

norling replied:
No, it's just a herd that didn't have the Herd type

danr commented on frontend/individuals.tsx:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Shows a list of all genebanks, with links to the individual genebanks.
+ */
+export function IndividualsTable({id}: {field: number}) {
```
Type should be `{id: number}`?

norling committed:
add yarn.lock

danr commented on frontend/manage_users.tsx:
```diff
@@ -0,0 +1,81 @@
+        onChange={userChange}
+        className={classes.verticalTabs}
+      >
+      {users.map((u:any, i:number) => <Tab key={i} label={u.email} />)}
```
u is a `NameID` which only has a `name` and `id` field, no `email` field

danr commented on frontend/manage_users.tsx:
```diff
@@ -0,0 +1,81 @@
+  const classes = useStyles()
+  const {users} = useDataContext()
+  const [userTab, setUserTab] = React.useState(0);
+  const [selectedUser, selectUser] = React.useState(undefined as number | string | undefined)
```
This should probably not be called `selectedUser` but something with edit action.

You can narrow down the string type to `'new'`: `number | 'new' | undefined`

danr commented on frontend/manage_user.tsx:
```diff
@@ -0,0 +1,220 @@
+
+  return <>
+    {user && <>
+      <h2>{isNew ? 'Ny användare' : user.email}</h2>
```
Move this and its state to an own `CreateNewUser` component

danr commented on frontend/manage_user.tsx:
```diff
@@ -0,0 +1,220 @@
+/**
+ * Provides a form for changing user metadata and user roles.
+ */
+export function ManageUser({id}: {id: number | string | undefined}) {
```
Here `id: number | 'new' | undefined` should probably not be called id...

Maybe split this into two components `ManageUser` and `CreateNewUser` and call the correct one from `ManageUsers`?

danr commented on frontend/manage_user.tsx:
```diff
@@ -0,0 +1,220 @@
+    if (genebankData.length > 0) {
+      setHerds(genebankData[0].herds)
+      if (herds.length > 0 && herds.filter(h => herd == h.id).length == 0) {
+        setHerd(herds[0].id)
```
Why do you first filter and then anyway set to the first elemnet?

norling replied:
it might be a dumb way to write this check, but it's essentially "if there are herds, and the currently selected herd isn't in the current list of herds, then select the first herd as default" could be `herds.length > 0 && !herds.find(h => h.id == herd)` instead, and that might be nicer...

danr replied:
Haha okej now I see it. I overlooked the `== 0` part but now that you pointed it out it's obvious of course. Yes `herds.length && !herd.find(...)` would be clearer. (Too bad `!` isn't `not` in JS though!)

Commit:
Use find instead of filter

Commit:
Add Icons type from MaterialTable

Commit:
Remove unnecessary types

Commit:
Fix typo longitud -> longitude

danr commented on frontend/manage_user.tsx:
```diff
@@ -0,0 +1,220 @@
+                      type='password'
+                      hidden={true}
+                      className={classes.simpleField}
+                      value={user.password ?? undefined}
```
This needs to be `''` instead of `undefined` otherwise React gets confused if this is a controlled or uncontrolled component.

norling replied:
ok! good to know! :)

danr commented on frontend/manage_users.tsx:
```diff
@@ -0,0 +1,81 @@
+// Define styles for tab menu
+const useStyles = makeStyles({
+  sidebar: {
+    float: "left",
```
Why not put sidebar and controls side-by-side using flex instead of float?

danr replied:
Then you don't need to do all the calc.

norling replied:
I honestly don't remember. I almost always use flex, so I want to think I had a reason for a float, but I can't remember what it would be :D I'll have a look through the rules once we have some more data on what they want from the thing though! I'll probably design everything like a big flex-thing in the end...

danr replied:
Alright!

danr commented on frontend/individuals.tsx:
```diff
@@ -0,0 +1,96 @@
+        columns={columns}
+        data={individuals}
+        title="Alla individer"
+      />
```
Moder och Fader seems to refer to some internal id (like 11793) while Nummer for a given bunny is written in another form (like G1390-1911).

norling replied:
yeah - this is purely a database optimization issue. I'd prefer to have parents as NameID or something like that, but I haven't had time to sort out a fast enough database query to get all the info I want.

danr replied:
Alright!

danr commented on frontend/manage.tsx:
```diff
@@ -0,0 +1,100 @@
+  const {genebanks} = useDataContext()
+
+  const tabs: ui.RoutedTab[] = genebanks.map(g => ({
+    path: `/${g.id}`,
```
What about 
```typescript
    path: `/${g.name.toLowerCase().replace(/åä/g, 'a').replace(/ö/g, 'o')}`,
```
?

norling replied:
links will probably be changed to the G-id later, so I'll keep this for now.

danr commented on frontend/explore_herds.tsx:
```diff
@@ -0,0 +1,89 @@
+// Define styles for tab menu
+const useStyles = makeStyles({
+  sidebar: {
+    float: "left",
```
Some interesting overflow is happening here
![image](https://user-images.githubusercontent.com/199380/85532425-fa666280-b60f-11ea-8ab0-1ebdebb42ad1.png)

danr commented on frontend/individuals.tsx:
```diff
@@ -0,0 +1,96 @@
+  },
+});
+
+const tableIcons = {
```
Move all the icons stuff to an own file?

norling replied:
Yeah - we can do that once the table is finished. In the mean time I'd prefer to keep them here so that I can look and change easier :)

danr commented on frontend/manage_herds.tsx:
```diff
@@ -0,0 +1,96 @@
+  }
+});
+
+export function ManageHerds({id}: {id: number | undefined}) {
```
There is a lot duplicated here and in `ExploreHerds`. Why not just add an edit button when Exploring if you're an admin?

norling replied:
yeah - it wasn't originally supposed to be so similar, but I ran out of time. so I'll have to think about the final design before I decide if there's value in optimizing.

danr replied:
One option is just to write this down in an issue and let it be like it is right now of course.

danr commented on frontend/data_context_global.tsx:
```diff
@@ -0,0 +1,84 @@
+    wire_phone_privacy?: string | null
+    latitude: string | null
+    longitude: string | null
+    coordinates_privacy?: string | null
```
Many fields here are both `undefine`-able (with `?`) and `| null`. Why both?

norling replied:
they're undefined when they're not present and null when they're unset.

danr replied:
Ah I get it, it's for the `_privacy` fields

danr reviewed (changes requested):
Comments from a (first?) run through the frontend. Also some accompanying changes in #83

norling committed:
Merge pull request #83 from NBISweden/danr/frontend-layout

Small fixes while reviewing

pontus committed:
Placeholder files for config directory

pontus committed:
No special development treatment

pontus committed:
New files for deployment

pontus committed:
Container build changes for deployment

pontus committed:
Changes to ease deployment

pontus committed:
Make sure development environment works more consistently

norling committed:
fix: css overflow

norling committed:
add herd_name to herd nmanagement form

pontus committed:
Compute backend URL correctly)

norling committed:
fix: herdForm id type

pontus committed:
Run registration action with the setup in container instead

pontus committed:
Disable secure cookies for now

norling committed:
nicer herdForm types

danr commented on frontend/herdForm.tsx:
```diff
@@ -51,9 +52,8 @@ export function HerdForm(props: {id: string | number | undefined}) {
-    herd[label] = value;
-    setHerd(herd);
+  const setFormField = <K extends keyof Herd>(label: K, value: Herd[K]) => {
+    herd && {...herd, label: value}
```
You lost `setHerd` :fearful:

danr replied:
And the brackets in `[label]`!

`setHerd({...herd, [label]: value})`

norling replied:
haha - yeah, that was dumb ^^

norling committed:
Better sorting for herd lists

norling committed:
Add config location via environment variable

norling committed:
remove special case backend url

pontus committed:
Use bash for entryscript to allow {a,b,c} usage

pontus committed:
Fix command for parcel server

pontus committed:
Copy configuration (if any) to temporary storage and use that

Currently, the WSGI code will try to provide a configuration if
none is provided. If we point the configuration environment variable
(HERDBOOK_CONFIG) to the /config directory, the server may not be able
to create it there for permissions reasons (or r/o).

Copy any configuration to /tmp and use that instead.

pontus committed:
Make sure wsgi server can read configuration

norling committed:
Merge pull request #84 from NBISweden/feature/frontend-layout-deploy

Changes to support easier deployment while maintaining a quick dev cycle

norling committed:
fix: missing setHerd in herdForm

norling committed:
type for postData variable in submitForm

norling committed:
Improved data for individuals table

norling committed:
More static info on temp herd page

norling committed:
fix: missing key in loop

norling committed:
fix: missing loop key

danr commented on frontend/herd_view.tsx:
```diff
@@ -28,6 +38,14 @@ export function HerdView({id}: {id: string}) {
       <>
         <h2>{`G${herd.herd}`} {herd.herd_name ? `- ${herd.herd_name}` : ''}</h2>
+        <dl>
+          {fields.map(field => <>
```
It's only after `return` you need `<> ... </>` or `( ... )`, not when you're in an unfinished expression such as to the right of `=>`.

danr replied:
Oh I didn't see you had both `<dt>` and `<dd>`, then you need a `<>`-fragment of course.

norling replied:
yeah, but I had to change it to a div to get a key on it, so I still did it wrong ^^

norling committed:
Add api date formatting

norling committed:
fix: add id to useEffect dependencies

norling committed:
Add link to individual from name field in table

norling committed:
Move link from name to number in table

jhagberg commented:
The Mellerud data is not listing 

/genebanks/2/individuals
and 
/genebanks/1/individuals

is the same list

norling committed:
fix: select genebank individuals by genebank id
